### PR TITLE
Replace download with git system

### PR DESCRIPTION
**Pull request by Chiiizzzy**

### CODE REVIEW

File: towhee/hub/file_manager.py [hunk 1]
```diff
@@ -228,12 +227,11 @@ def _cache_name(self, name: str, author: str, branch: str):
...
             file_name, file_type = name.split('.')
         except ValueError:
-            return name.replace('-', '_') + '&' + author + '$' + branch
-
+            return author + '/' + name.replace('-', '_') + '/' + branch
         # If pointing directly to pipeline or operator file.
         if file_type in ['py', 'yaml']:
-            return file_name.replace('-', '_') + '&' + author + '$' + branch
-        # If not pointing to a pipeline or operator file.
+            return author + '/' + file_name.replace('-', '_') + '/' + branch
```
Comment (junjiejiangjjj) [inline towhee/hub/file_manager.py:233]: We should use `pathlib` to join path, for `towhee` need to support windows, Linux and Mac.

File: towhee/hub/file_manager.py [hunk 2]
```diff
@@ -280,7 +278,7 @@ def _cache_locals(self):
                     # TODO: Figure out exception types
                     copytree(str(old_path), str(new_path))
 
-    def get_pipeline(self, pipeline: str, branch: str = 'main', redownload: bool = False, install_reqs=True):
+    def get_pipeline(self, pipeline: str, branch: str = 'main', install_reqs=True):
```
Comment (junjiejiangjjj) [inline towhee/hub/file_manager.py:281]: We should change the name from `branch` to `tag`,  and the default value is `None`, which means the pipeline will use `main`.

File: towhee/hub/file_manager.py [hunk 3]
```diff
@@ -329,22 +325,24 @@ def get_pipeline(self, pipeline: str, branch: str = 'main', redownload: bool = F
...
+                repo = git.Repo(repo_path)
+                if not 'Your branch is up to date' in repo.git.status():
+                    print(f'Your local pipeline {repo_name} is not up to date, updating to latest version...')
+                    fetch_head = repo.remote('origin').fetch(branch)
+                    repo.git.reset(fetch_head, '--hard')
 
         return file_path
 
-    def get_operator(self, operator: str, branch: str = 'main', redownload: bool = False, install_reqs=True):
+    def get_operator(self, operator: str, branch: str = 'main', install_reqs=True):
```
Comment (junjiejiangjjj) [inline towhee/hub/file_manager.py:345]: The same with the previous comment.

File: towhee/hub/file_manager.py [hunk 4]
```diff
@@ -393,16 +389,19 @@ def get_operator(self, operator: str, branch: str = 'main', redownload: bool = F
...
+                return file_path
+
+            if file_path.is_file() is False:
+                download_repo(author, repo, branch, str(file_path.parent), install_reqs=install_reqs)
             else:
-                if redownload:
-                    # TODO: filip-halt
-                    # Error logging.
-                    rmtree(file_path.parent)
+                repo_path = str(path).rsplit('/', 1)[0]
```
Comment (junjiejiangjjj) [inline towhee/hub/file_manager.py:399]: Use `pathlib`

File: towhee/hub/file_manager.py [hunk 5]
```diff
@@ -393,16 +389,19 @@ def get_operator(self, operator: str, branch: str = 'main', redownload: bool = F
...
+                repo_name = str(path).rsplit('/', 1)[1].split('.')[0]
+                repo = git.Repo(repo_path)
+                if not 'Your branch is up to date' in repo.git.status():
+                    print(f'Your local operator {repo_name} is not up to date, updating to latest version...')
```
Comment (junjiejiangjjj) [inline towhee/hub/file_manager.py:403]: Use `engine_log`

File: towhee/hub/file_manager.py [hunk 6]
```diff
@@ -393,16 +389,19 @@ def get_operator(self, operator: str, branch: str = 'main', redownload: bool = F
...
+                    fetch_head = repo.remote('origin').fetch(branch)
+                    repo.git.reset(fetch_head, '--hard')
```
Comment (junjiejiangjjj) [inline towhee/hub/file_manager.py:405]: Why use `reset`, if having some changes in local files, they will be lost.

File: towhee/hub/file_manager.py (hunk 1 above)
Comment (Chiiizzzy) [reply towhee/hub/file_manager.py:233]: fixed

File: towhee/hub/file_manager.py (hunk 2 above)
Comment (Chiiizzzy) [reply towhee/hub/file_manager.py:281]: Will create another PR to change all the `branch` to `tag`

File: towhee/hub/file_manager.py (hunk 3 above)
Comment (Chiiizzzy) [reply towhee/hub/file_manager.py:345]: Will create another PR to change all the `branch` to `tag`

File: towhee/hub/file_manager.py (hunk 4 above)
Comment (Chiiizzzy) [reply towhee/hub/file_manager.py:399]: fixed

File: towhee/hub/file_manager.py (hunk 5 above)
Comment (Chiiizzzy) [reply towhee/hub/file_manager.py:403]: fixed

File: towhee/hub/file_manager.py (hunk 6 above)
Comment (Chiiizzzy) [reply towhee/hub/file_manager.py]: fixed

**Pull request merged**